File: frontend/src/api/recognition.ts
/**
 * api/recognition.ts
 * 通用识别 + 历史记录 + 反馈
 */
import request from './request';

export type RecognitionType = 'travel' | 'ocr' | 'currency' | 'product' | 'face';

export interface RecognitionParams {
  type: RecognitionType;
  image?: Blob | File;
  extra?: Record<string, any>;
}

export interface RecognitionResult {
  id: string;
  type: RecognitionType;
  title: string;
  content: string;
  audioUrl?: string;
  thumbnail?: string;
  createdAt: string;
}

/**
 * WebRTC 实时帧识别结果（与 RecognitionResult 兼容，并携带额外字段）
 */
export interface LiveRecognitionResult extends RecognitionResult {
  priority?: 'high' | 'normal';
  framesUsed?: number;
  latencyMs?: number;
  resultJson?: {
    confidence?: number;
    is_demo?: boolean;
    is_fallback?: boolean;
    priority?: 'high' | 'normal';
    frames_used?: number;
    mode?: string;
  };
}

export interface HistoryParams {
  page?: number;
  pageSize?: number;
  type?: RecognitionType;
  keyword?: string;
}

export interface HistoryListResp {
  list: RecognitionResult[];
  total: number;
  page: number;
  pageSize: number;
}

export function recognize(params: RecognitionParams) {
  const form = new FormData();
  if (params.image) form.append('image', params.image);
  if (params.extra) form.append('extra', JSON.stringify(params.extra));
  return request.post<any, RecognitionResult>(`/recognition/${params.type}`, form, {
    headers: { 'Content-Type': 'multipart/form-data' },
  });
}

/**
 * WebRTC 实时帧识别
 *
 * 由前端 useLiveStream 收集到 N 张 base64 帧图后批量调用。
 *  - 1 ~ 8 张 base64 字符串，可含 "data:image/jpeg;base64," 前缀（后端会自动剥离）
 *  - 后端走 JoyAI-VL-Interaction 的视频流多帧推理
 *  - 返回中含 priority=high 表示命中危险关键词（视障场景里要立刻播报）
 */
export function recognizeLive(payload: {
  scene: RecognitionType;
  images: string[];
  extra?: Record<string, any>;
  saveRecord?: boolean;
}) {
  return request.post<any, LiveRecognitionResult>('/recognition/live', payload);
}

export function getHistory(params: HistoryParams) {
  return request.get<any, HistoryListResp>('/recognition/history', { params });
}

export function getHistoryDetail(id: string) {
  return request.get<any, RecognitionResult>(`/recognition/history/${id}`);
}

export function deleteHistory(id: string) {
  return request.delete(`/recognition/history/${id}`);
}

export function submitFeedback(payload: { id: string; rating: number; comment?: string }) {
  return request.post('/recognition/feedback', payload);
}

File: backend/app/schemas/recognition.py
"""
识别业务相关
"""
from __future__ import annotations

from datetime import datetime
from typing import Any, List, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field

RecognitionType = Literal["travel", "ocr", "currency", "product", "face"]


class RecognitionResult(BaseModel):
    """前端契约：识别结果（camelCase）"""

    id: str
    type: RecognitionType
    title: str
    content: str
    audioUrl: Optional[str] = None
    thumbnail: Optional[str] = None
    createdAt: str
    # 扩展字段（与前端 useRecognition 兼容）
    resultJson: Optional[dict] = None
    confidence: Optional[float] = None
    isFallback: Optional[bool] = None


class HistoryListResp(BaseModel):
    """前端契约：list/total/page/pageSize"""

    list: List[RecognitionResult]
    total: int
    page: int
    pageSize: int


class HistoryQuery(BaseModel):
    page: int = 1
    pageSize: int = 20
    type: Optional[RecognitionType] = None
    keyword: Optional[str] = None


class FeedbackSubmitReq(BaseModel):
    """识别结果反馈"""

    id: str = Field(..., description="识别记录 ID")
    rating: int = Field(..., ge=1, le=5, description="评分 1-5")
    comment: Optional[str] = Field(default=None, max_length=500)


class LiveRecognitionReq(BaseModel):
    """
    WebRTC 实时帧识别请求

    前端通过 getUserMedia 拿到视频流后，定期截帧转 base64，
    按"批"调用本接口（推荐 4 帧/批，对应 4 秒 1 帧）。
    后端会把多帧一次性喂给 JoyAI-VL-Interaction（视频流模式），
    返回结构化场景描述 + 危险关键词优先级。
    """

    scene: RecognitionType = Field(default="travel", description="业务场景")
    # 1~8 张 base64 编码的 jpeg 图片（不含 "data:image/jpeg;base64," 前缀也可）
    images: List[str] = Field(..., min_length=1, max_length=8, description="base64 帧图列表")
    extra: Optional[dict[str, Any]] = Field(default=None, description="Prompt 模板变量")
    # 是否把这一批结果落库为一条识别记录（默认 true，便于历史回放）
    saveRecord: Optional[bool] = Field(default=True, description="是否保存到 rec_record")

